packages/integration-sdk-cli/src/neo4j/neo4jGraphStore.ts

```diff
@@ -0,0 +1,135 @@
+import { Entity, Relationship } from '@jupiterone/integration-sdk-core';
+
+import * as neo4j from 'neo4j-driver';
+
+export interface Neo4jGraphObjectStoreParams {
+  uri: string;
+  username: string;
+  password: string;
+}
+
+export class Neo4jGraphStore {
+  private neo4jDriver: neo4j.Driver;
+  private persistedSession: neo4j.Session;
+  private databaseName = 'neo4j';
+  private typeList = new Set<string>();
+
+  constructor(params: Neo4jGraphObjectStoreParams, session?: neo4j.Session) {
+    if(session) {
+      this.persistedSession = session;
+    }
+    else {
+      this.neo4jDriver = neo4j.driver(
+        params.uri,
+        neo4j.auth.basic(params.username, params.password),
+      );
+    }
+  }
+
+  private async runCypherCommand(cypherCommand: string): Promise<neo4j.Result> {
+    if(this.persistedSession) {
+      const result = await this.persistedSession.run(cypherCommand);
+      return result;
+    }
+    else {
+      const session = this.neo4jDriver.session({
+        database: this.databaseName,
+        defaultAccessMode: neo4j.session.WRITE,
+      });
+      const result = await session.run(cypherCommand);
+      await session.close();
+      return result;
+    }
+  }
+
+  private startsWithNumeric(str: string){
+    return /^\d+$/.test(str);
+  }
+
+  private sanitizePropertyName(propertyName: string): string {
+    let sanitizedName = '';
+    if(this.startsWithNumeric(propertyName)) {
+      sanitizedName += 'n';
+    }
+    sanitizedName += propertyName;
+    sanitizedName = sanitizedName.replace(/[\s!@#$%^&*()-=+\\|'";:/?.,><`~\t\n[\]{}]/g, "_");
+    return sanitizedName;
+  }
+
+  private buildPropertyString(propList, nodeName) {
+    let propString = '';
+    for (const key in propList) {
+      if (key === '_rawData') {
+        //stringify JSON in rawData so we can store it.
+        propString += `${nodeName}.${key} = '${JSON.stringify(propList[key])}', `;
+      } else {
+        // Escape single quotes so they don't terminate strings prematurely
+        const finalValue = propList[key].toString().replace(/'/gi, "\\'");
+        // Sanitize out characters that aren't allowed in property names
+        const propertyName = this.sanitizePropertyName(key);
+        propString += `${nodeName}.${propertyName} = '${finalValue}', `;
+      }
+    }
+    propString = propString.slice(0, -2);
+
+    return propString;
+  }
+
+  async addEntities(newEntities: Entity[]) {
+    for await (const entity of newEntities) {
+      //Add constraint if not already in types
+      //We check for existence in case we're ever running in an instance with
+      //multiple inputs.
+      if (!this.typeList.has(entity._type)) {
+        await this.runCypherCommand(
+          `CREATE CONSTRAINT unique_${entity._type} IF NOT EXISTS ON (n:${entity._type}) ASSERT n._key IS UNIQUE;`,
```

Comment:
I've updated this to be an INDEX instead of a constraint.  If we want a constraint in across all nodes, we'll need to do something like add a shared label to all nodes and create a constraint on that.  But I've left that out for now as it is likely to cause some visualization quirks that I don't think we want.  Can address later if we run into issues.

Reply:
Sounds good. I know trying to map intricacies of J1 to the already-established patterns of Neo4j was probably a bit of a head scratcher, but I think you've done an excellent job in the decisions you've made for this CLI. Thank you again for all of your work on this!